### PR TITLE
feat(studio): AST details

### PR DESCRIPTION
**Pull request by grzim**
image

### CODE REVIEW

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx [hunk 1]
```diff
@@ -122,15 +196,18 @@ export const ASTViewer: React.FC<Props> = ({ type }) => {
     );
   }
 
-  const transformedRootNode = transformTreeData(rootNode);
+  const transformedRootNode =
+    transformTreeData(mapBabelASTToRenderableTree(rootNode)) || null;
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:200]: please use nullish-coalescing `??`

File: apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
+export function extractComments(node: Node): [Node, Node[]] {
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:1]: let's return an object with named properties here

File: apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        } else if (typeof value === "object" && value !== null) {
+          const traversed = traverse(value, node);
+          if (traversed !== undefined) {
+            result[key] = traversed;
+          }
+        } else {
+          result[key] = value;
+        }
+      }
+    }
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:40]: reduce with early returns would help avoiding 3 conditional statements and one could argue it's easier to read that way. if you agree, let's switch

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx [hunk 4]
```diff
@@ -27,12 +29,85 @@ interface Props {
...
+        const isParam = paramsIds.includes(getPosition(child));
+        const nameMap = {
+          BlockStatement: "body",
+          Identifier: "id",
+        };
+        return transformTreeData(
+          child,
+          (isParam ? "param" : nameMap[child.label]) || child.label,
+        );
+      }) || [];
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:84]: use nullish coalescing `??`

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx [hunk 5]
```diff
@@ -27,12 +29,85 @@ interface Props {
...
+      node.children?.map((child) =>
+        transformTreeData(
+          child,
+          Array.isArray(node.actualNode[node.label])
+            ? node.label
+            : Object.keys(node.actualNode).find(
+                (key) => node.actualNode[key] === child.actualNode,
+              ),
+        ),
+      ) || [];
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:96]: use nullish coalescing `??`

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx [hunk 6]
```diff
@@ -27,12 +29,85 @@ interface Props {
...
+
+  if (node.actualNode.type === "CallExpression") {
+    const args = node.actualNode.arguments.map(parseNodeData(node, "arg"));
+    const argumentsIds = args.map(getPosition);
+    result.children =
+      node.children.map((child) => {
+        const isParam = argumentsIds.includes(getPosition(child));
+        return transformTreeData(child, isParam ? "argument" : child.label);
+      }) || [];
+  } else if (node.actualNode.type === "FunctionDeclaration") {
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:68]: prefer early returns over nested conditionals if it's possible to use them please

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx [hunk 7]
```diff
@@ -27,12 +29,85 @@ interface Props {
...
+        const nameMap = {
+          BlockStatement: "body",
+          Identifier: "id",
+        };
+        return transformTreeData(
+          child,
+          (isParam ? "param" : nameMap[child.label]) || child.label,
+        );
+      }) || [];
+  } else {
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:85]: prefer early returns over nested conditionals if it's possible to use them please

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx (hunk 6 above)
Comment (grzim) [reply apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:68]: @r4zendev what do you mean by nested if? We have just one level of nesting here (it is else if, not if within an if)

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx (hunk 7 above)
Comment (grzim) [reply apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:85]: same as above - we dont have nested conditions - we need to check the node type to make correct mapping (and add relations)

File: apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx (hunk 6 above)
Comment (r4zendev) [reply apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx:68]: yeah, i phrased it incorrectly. in general, early returns are easier to read than multiple `else if` statements. here, on line 99 you return result that is being mutated in each conditional. the alternative is to return result on lines 68, 85, 97 and replace `else if` with `if`, remove `else`. i could argue that this makes the code more readable. if you agree, let's replace it with early returns. i agree with both, but prefer my option slightly more.

Comment (grzim) [reply apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx]: Updates added. When it comes to `comments` I made it mutable as we are traversing the tree so it simplifies the code a lot this way

File: apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      }
+      return { ...acc, [key]: value };
+    }, {});
+  }
+
+  const cleanedStructure = traverse(node);
+  return { cleanedStructure, comments } as {
+    cleanedStructure: Node;
+    comments: Node[];
+  };
```
[the remark this answers is not in the document]
Comment (grzim) [reply apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:47]: due to all the traversal TS is loosing its mind here, somewhere I will need to add a casting anyway. Will think about it

File: apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        const mapped = value
+          .map((item) => traverse(item, node))
+          .filter((item) => item !== undefined);
+        if (mapped.length > 0) return { ...acc, [key]: mapped };
+      }
+      if (typeof value === "object" && value !== null) {
+        const traversed = traverse(value, node);
+        return traversed ? { ...acc, [key]: traversed } : acc;
+      }
+      return { ...acc, [key]: value };
```
[the remark this answers is not in the document]
Comment (grzim) [reply apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:39]: I will be bold and even add new lines

File: apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (grzim) [reply apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:47]: I found some functionality issues to fix (I converted PR to draft) will focus on types when doing fixes

**Comment by grzim** [comment]
image

File: apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  return traverse(root);
+}
+
+export const removeEmptyChildren = (node: TreeNode): TreeNode => {
+  const transformedChildren =
+    node.children
+      ?.map(removeEmptyChildren)
+      .filter((child) => child.children?.length || !child.children) || [];
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:45]: `??`

File: apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  return {
+    ...node,
+    children: transformedChildren.length > 0 ? transformedChildren : undefined,
+  };
+};
+
+export const transformNode = (content: string, type: EditorType) => {
+  const parsed = parseSnippet(content);
+  const rootNode = isFile(parsed) ? mapBabelASTToRenderableTree(parsed) : null;
+  if (!rootNode) return null;
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:56]: please wrap in a block statement `{ }`

Comment (r4zendev) [reply apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts:56]: if possible, also add spaces between logically groupped code lines

File: apps/frontend/app/(website)/studio/src/store/snippets.ts [hunk 12]
```diff
@@ -108,174 +107,186 @@ const getEditorsFromLS = () => {
...
+  tabsLimit: 12,
+  getHasReachedTabsLimit: () => get().editors.length >= get().tabsLimit,
+  // editors: localStorage.getItem('editors') ? JSON.parse(localStorage.getItem('editors')) : [getSingleTestCase()],
+  addPair: (name?: string) => {
+    set({
+      editors: [
+        ...get().editors,
+        {
+          name:
+            name ||
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/src/store/snippets.ts:120]: `??`

File: apps/frontend/app/(website)/studio/src/store/snippets.ts [hunk 13]
```diff
@@ -108,174 +107,186 @@ const getEditorsFromLS = () => {
...
+      set({
+        selectedPairIndex: 0,
+        editors,
+      });
+    } else
+      set({
+        selectedPairIndex: Math.max(get().selectedPairIndex - 1, 0),
+        editors,
+      });
+  },
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/src/store/snippets.ts:165]: spaces and early returns please

File: apps/frontend/app/(website)/studio/src/store/snippets.ts [hunk 14]
```diff
@@ -108,174 +107,186 @@ const getEditorsFromLS = () => {
...
+          after: [],
+          output: [],
+        } as AllEditors,
+        get().editors,
+      ),
+    );
+  },
+  setSelectedPairIndex: (i: number) => {
+    set({ selectedPairIndex: i });
+  },
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/src/store/snippets.ts:207]: use shorthand return as on line 186

File: apps/frontend/app/(website)/studio/src/store/snippets.ts [hunk 15]
```diff
@@ -108,174 +107,186 @@ const getEditorsFromLS = () => {
...
+      setBeforeSelection: get().setSelection(index, "before"),
+      setAfterSelection: get().setSelection(index, "after"),
+      setOutputSelection: get().setSelection(index, "output"),
+      setSelection: (editorType: EditorType) =>
+        get().setSelection(index, editorType),
+    };
+  },
+  setInitialState: (state) => {
+    set(state);
+  },
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/src/store/snippets.ts:229]: shorthand return

File: apps/frontend/app/(website)/studio/src/store/snippets.ts [hunk 16]
```diff
@@ -108,174 +107,186 @@ const getEditorsFromLS = () => {
...
+        localStorage.setItem(
+          "editors",
+          JSON.stringify(obj.editors.map(toEditorSnippets)),
+        );
+      } catch (error) {
+        console.error("error on JSON.stringify(obj.editors) ", { error });
+      }
+    };
+  },
+  setEditors: (editorsContents: EditorsSnippets[]) => {
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/src/store/snippets.ts:255]: shorthand return

File: apps/frontend/app/(website)/studio/src/store/snippets.ts (hunk 14 above)
Comment (r4zendev) [reply apps/frontend/app/(website)/studio/src/store/snippets.ts:207]: or use block statements everywhere, i'm ok with both, just has to be consistent

File: apps/frontend/app/(website)/studio/src/utils/babelParser/index.ts [hunk 17]
```diff
@@ -12,7 +12,7 @@ const isParseError = (err: unknown): err is ParseError =>
   Object.hasOwn(err, "reasonCode");
 
 const parseSnippet = (snippet: string) => {
-  if (!snippet || snippet.trim() === "") {
+  if (!snippet || snippet.trim?.() === "") {
```
Comment (r4zendev) [inline apps/frontend/app/(website)/studio/src/utils/babelParser/index.ts:15]: make the type of snippet on line 14 string | undefined if that's the case.

File: apps/frontend/app/(website)/studio/src/store/snippets.ts (hunk 15 above)
Comment (grzim) [reply apps/frontend/app/(website)/studio/src/store/snippets.ts:229]: ????

File: apps/frontend/app/(website)/studio/src/store/snippets.ts (hunk 16 above)
Comment (grzim) [reply apps/frontend/app/(website)/studio/src/store/snippets.ts:255]: no idea what do you mean - it is a store

File: apps/frontend/app/(website)/studio/src/store/snippets.ts (hunk 12 above)
Comment (grzim) [reply apps/frontend/app/(website)/studio/src/store/snippets.ts:120]: nope - we dont want to have empty string here

File: apps/frontend/app/(website)/studio/src/store/snippets.ts (hunk 13 above)
Comment (grzim) [reply apps/frontend/app/(website)/studio/src/store/snippets.ts:165]: this one has not been touched in this PR

**Pull request merged**